### PR TITLE
Add AdaLayerNormZero normalization module for Diffusion Transformers

### LINKED ISSUE #2796: Feature Request: Video Generation Model Support (Wan, DiT architectures)
## Summary

Request support for video generation models in Megatron Core, specifically Diffusion Transformer (DiT) based architectures like Alibaba's Wan, with optimizations from Transformer Engine including custom FP8 attention kernels.

## Motivation

Video generation is a rapidly growing area with models like Wan 2.1, CogVideoX, and OpenSora demonstrating impressive results. However:

- **Incomplete DiT architecture support** — Megatron Core has some DiT foundations (conditional embedder gradient sync, bidirectional attention) but lacks key components like adaLN-Zero conditioning
- **Missing video-specific optimizations** — Video generation requires handling 3D spatiotemporal tensors with specialized 3D patch embeddings
- **Transformer Engine FP8 untapped** — TE has custom FP8 attention kernels (`fp8_dpa`, `fp8_mha`) that could significantly accelerate video DiT training but aren't integrated for this use case
- **No reference implementation** — While infrastructure exists, there's no end-to-end video generation model example

## Current State

Megatron Core has relevant foundations but no video generation support:
- **MIMO framework** (`megatron/core/models/mimo/`) — Experimental multimodal with video *understanding* (Video-LLaVA), not generation
- **Transformer Engine integration** — `TEDotProductAttention` with FP8 support, but not adapted for DiT architectures
- **Energon dataloader** — Video decoding support exists for understanding tasks

## Ask

1. **Adaptive LayerNorm (adaLN-Zero)** — Implement the core DiT conditioning

2. **3D patch embedding for video** — Extend vision patch embedding to handle temporal dimension

3. **Wan 2.1 reference implementation** — End-to-end example with:
   - Text-to-video generation
   - Image-to-video generation  
   - Training scripts with distributed parallelism

4. **Transformer Engine FP8 optimization for DiT** — Validate and document FP8 training:
   - `fp8_dpa` for video self-attention

## References

- [Wan 2.1](https://github.com/Wan-Video/Wan2.1) — Alibaba's open video generation model
- [DiT Paper](https://arxiv.org/abs/2212.09748) — Scalable Diffusion Models with Transformers
- https://github.com/NVIDIA/Megatron-LM/issues/1592

### PR DESCRIPTION
**Pull request by CodersAcademy006**
This PR adds an implementation of **Adaptive LayerNorm Zero (adaLN-Zero)**, a core building block required for Diffusion Transformer (DiT) architectures.

The module follows the DiT specification exactly:
- Produces modulation parameters (gamma, beta)
- Produces a **gate (alpha)** initialized to zero
- Returns `(x_modulated, gate)` without applying residuals or compute

This preserves identity behavior at initialization and allows correct gating **after** Attention / MLP layers in the parent Transformer block.

### What is included

- `AdaLayerNormZero` module under `megatron/core/transformer/`
- Zero-initialized conditioning projection (DiT-style stability)
- Explicit typing for constructor and forward return values
- Unit tests covering:
  - Identity behavior at initialization
  - Conditioning effect when weights are non-zero
  - Broadcasting from `[B, C] → [B, S, C]`

Megatron Core already contains partial DiT foundations (bidirectional attention, multimodal infrastructure), but lacks adaLN-Zero, which is required by modern DiT-based diffusion models (e.g., Wan, DiT, OpenSora).

This change completes that missing normalization primitive while preserving existing architecture boundaries.

### Tests

```bash
pytest tests/unit_tests/transformer/test_adaptive_layernorm.py
```
Part 1- Fixes #2796

### CODE REVIEW

**Comment by CodersAcademy006** [comment]
@sbhavani Hey, once free please feel free to review this and provide me relevant suggestions to modify any changes needed. Thank You.

**Comment by Phlip79** [comment]
We plan to support DiT soon in Megatron. Closing this PR. Also, if you're a bot you should make that clear in your name.

**Comment by CodersAcademy006** [comment]
@Phlip79 Thanks for the clarification.

Glad to hear DiT support is planned for Megatron looking forward to that. I'll keep an eye on the upcoming implementation.

If there are other areas where external contributions would be useful for the diffusion/video stack, happy to help.

Thank you for your response.
